### ISSUE TITLE
UCSF VPN works with OpenConnect (>= 7.08) [WARNING!!!]

### ISSUE THREAD

**Issue by HenrikBengtsson** [HenrikBengtsson/ucsf-vpn#4]
**WARNING WARNING WARNING: If you attempt the below OpenConnect 7.08 hack on Ubuntu 16.04, you might end up with a broken system.  Proceed on your own risk.**

# Background
In mid-August 2017, OpenConnect stopped working with the UCSF VPN servers (which had been updated).  The observed error [reported](https://listsrv.ucsf.edu/cgi-bin/wa?A2=IT-FORUM;9377b28f.1708) was:
```sh
$ sudo openconnect --no-cert-check --user="$MYID" --authgroup="Single-Factor Pulse Clients" --juniper 'https://remote.ucsf.edu/pulse'
WARNING: Juniper Network Connect support is experimental.
It will probably be superseded by Junos Pulse support.
[...]
Connected to HTTPS on remote.ucsf.edu
Got HTTP response: HTTP/1.1 400 Bad Request
Unexpected 400 result from server
Creating SSL connection failed
```

# Troubleshooting / Solution
On 2017-10-09, @graft [found](https://listsrv.ucsf.edu/cgi-bin/wa?A2=IT-FORUM;dfc383f9.1710) that this is a [bug in OpenConnect (< 7.08)](https://serverfault.com/questions/835918/openconnect-and-pulse-stopped-working) and that it has been fixed on OpenConnect 7.08.  He reports that he can successful connect using:
```sh
$ sudo openconnect --servercert=sha256:82a0...<not sure if safe to share>...a304bb --authgroup="Single-Factor Pulse Clients" --juniper https://remote.ucsf.edu/openconnect
```
on Ubuntu 16.04 (Xenial).  To install OpenConnect 7.08 he had to do some tricks (see comments below).

**Comment by HenrikBengtsson** [comment]
# Installing OpenConnect 7.08 on Ubuntu 16.04 (Xenial)

Most people on Ubuntu are on 16.04 (Xenial) which only [provides OpenConnect 7.06](https://packages.ubuntu.com/xenial/net/openconnect):
```sh
$ openconnect --version
OpenConnect version v7.06
Using GnuTLS. Features present: PKCS#11, RSA software token, HOTP software token, TOTP software token, System keys, DTLS
```
whereas Ubuntu 17.04 (Zesty) [provides OpenConnect 7.08](https://packages.ubuntu.com/zesty/net/openconnect).   Unfortunately, OpenConnect is not among backported packages in [Xenial Backports](https://packages.ubuntu.com/xenial-backports/allpackages).


## Installation hack by Saurabh
On Ubuntu 16.04, @graft [reported success](https://listsrv.ucsf.edu/cgi-bin/wa?A2=IT-FORUM;dfc383f9.1710) after installing OpenConnect 7.08 binaries and dependencies from zesty/universe.  

**WARNING: It may corrupt your Ubuntu 16.04 installation if you install from 17.04.  I have not yet tried this myself.**

AFAIK, this was done by temporarily adding:
```
deb http://us.archive.ubuntu.com/ubuntu/ zesty universe
```
to `/etc/apt/sources.list`, running `sudo apt update` such that:
```sh
$ apt show openconnect
Package: openconnect
Version: 7.08-1
Priority: optional
Section: universe/net
Origin: Ubuntu
Maintainer: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
Original-Maintainer: Mike Miller <mtmiller@debian.org>
Bugs: https://bugs.launchpad.net/ubuntu/+filebug
Installed-Size: 2,336 kB
Depends: vpnc-scripts, libc6 (>= 2.14), libgnutls30 (>= 3.5.0), libopenconnect5 (>= 7.08), libproxy1v5 (>= 0.4.13), libxml2 (>= 2.7.4)
Homepage: http://www.infradead.org/openconnect/
Download-Size: 392 kB
APT-Sources: http://us.archive.ubuntu.com/ubuntu zesty/universe amd64 Packages
Description: open client for Cisco AnyConnect VPN
 OpenConnect is a client for Cisco's AnyConnect SSL VPN, which is supported by
 the ASA5500 Series, by IOS 12.4(9)T or later on Cisco SR500, 870, 880, 1800,
 2800, 3800, 7200 Series and Cisco 7301 Routers, and probably others.

N: There is 1 additional record. Please use the '-a' switch to see it
```

Then trying to install, we get:
```sh
$ sudo apt install openconnect
[...]
The following packages have unmet dependencies:
 openconnect : Depends: libgnutls30 (>= 3.5.0) but 3.4.10-4ubuntu1.4 is to be installed
               Depends: libopenconnect5 (>= 7.08) but 7.06-2build2 is to be installed
               Depends: libproxy1v5 (>= 0.4.13) but 0.4.11-5ubuntu1 is to be installed
E: Unable to correct problems, you have held broken packages.
```

@graft, is the above what you did and observed? How did you proceed from this stage?  Did you force install it somehow?

**Comment by graft** [comment]
You might need zesty main to get those packages updated.

Also, the --servercert argument seems to be unnecessary at all now - if you exclude it then openconnect will do a normal TLS handshake and validates the server certs it receives by conventional means.

**Comment by HenrikBengtsson** [comment]
So also adding
```
deb http://us.archive.ubuntu.com/ubuntu/ zesty main
```
did make installation of libopenconnect5 to work:
```sh
$ sudo apt install libopenconnect5
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Suggested packages:
  krb5-doc gnutls-doc gnutls-bin gnutls-bin:i386 krb5-user krb5-doc:i386 krb5-user:i386
Recommended packages:
  libtasn1-doc
The following packages will be REMOVED:
  libgnutls-dev
The following NEW packages will be installed:
  libkadm5clnt-mit11 libkadm5srv-mit11
The following packages will be upgraded:
  krb5-multidev libgnutls-openssl27 libgnutls28-dev libgnutls30 libgnutls30:i386 libgnutlsxx28
  libgssapi-krb5-2 libgssapi-krb5-2:i386 libgssrpc4 libk5crypto3 libk5crypto3:i386 libkrb5-3 libkrb5-3:i386
  libkrb5support0 libkrb5support0:i386 libopenconnect5 libproxy1-plugin-gsettings
  libproxy1-plugin-gsettings:i386 libproxy1-plugin-networkmanager libproxy1-plugin-webkit:i386 libproxy1v5
  libproxy1v5:i386 libtasn1-6 libtasn1-6:i386 libtasn1-6-dev
25 upgraded, 2 newly installed, 1 to remove and 2598 not upgraded.
Need to get 3,686 kB of archives.
After this operation, 1,450 kB of additional disk space will be used.

Do you want to continue? [Y/n] 
[...]
Fetched 3,686 kB in 13s (272 kB/s)                                                                             
(Reading database ... 558786 files and directories currently installed.)
Preparing to unpack .../libtasn1-6-dev_4.10-1_amd64.deb ...
Unpacking libtasn1-6-dev:amd64 (4.10-1) over (4.7-3ubuntu0.16.04.2) ...
Preparing to unpack .../libtasn1-6_4.10-1_i386.deb ...
De-configuring libtasn1-6:amd64 (4.7-3ubuntu0.16.04.2) ...
Unpacking libtasn1-6:i386 (4.10-1) over (4.7-3ubuntu0.16.04.2) ...
Preparing to unpack .../libtasn1-6_4.10-1_amd64.deb ...
Unpacking libtasn1-6:amd64 (4.10-1) over (4.7-3ubuntu0.16.04.2) ...
Processing triggers for man-db (2.7.5-1) ...
Processing triggers for libc-bin (2.23-0ubuntu9) ...
dpkg: libgnutls-dev:amd64: dependency problems, but removing anyway as you requested:
 libgnutls28-dev:amd64 depends on libgnutls-dev (= 3.4.10-4ubuntu1.4).
(Reading database ... 558784 files and directories currently installed.)
Removing libgnutls-dev:amd64 (3.4.10-4ubuntu1.4) ...
(Reading database ... 558756 files and directories currently installed.)
Preparing to unpack .../libgnutls28-dev_3.5.6-4ubuntu4_amd64.deb ...
Unpacking libgnutls28-dev:amd64 (3.5.6-4ubuntu4) over (3.4.10-4ubuntu1.4) ...
Preparing to unpack .../libgnutlsxx28_3.5.6-4ubuntu4_amd64.deb ...
Unpacking libgnutlsxx28:amd64 (3.5.6-4ubuntu4) over (3.4.10-4ubuntu1.4) ...
Preparing to unpack .../libgnutls-openssl27_3.5.6-4ubuntu4_amd64.deb ...
Unpacking libgnutls-openssl27:amd64 (3.5.6-4ubuntu4) over (3.4.10-4ubuntu1.4) ...
Preparing to unpack .../libgnutls30_3.5.6-4ubuntu4_amd64.deb ...
De-configuring libgnutls30:i386 (3.4.10-4ubuntu1.4) ...
Unpacking libgnutls30:amd64 (3.5.6-4ubuntu4) over (3.4.10-4ubuntu1.4) ...
Preparing to unpack .../libgnutls30_3.5.6-4ubuntu4_i386.deb ...
Unpacking libgnutls30:i386 (3.5.6-4ubuntu4) over (3.4.10-4ubuntu1.4) ...
Preparing to unpack .../krb5-multidev_1.15-1_amd64.deb ...
Unpacking krb5-multidev (1.15-1) over (1.13.2+dfsg-5ubuntu2) ...
Preparing to unpack .../libgssrpc4_1.15-1_amd64.deb ...
Unpacking libgssrpc4:amd64 (1.15-1) over (1.13.2+dfsg-5ubuntu2) ...
Selecting previously unselected package libkadm5srv-mit11:amd64.
Preparing to unpack .../libkadm5srv-mit11_1.15-1_amd64.deb ...
Unpacking libkadm5srv-mit11:amd64 (1.15-1) ...
Selecting previously unselected package libkadm5clnt-mit11:amd64.
Preparing to unpack .../libkadm5clnt-mit11_1.15-1_amd64.deb ...
Unpacking libkadm5clnt-mit11:amd64 (1.15-1) ...
Preparing to unpack .../libgssapi-krb5-2_1.15-1_amd64.deb ...
De-configuring libgssapi-krb5-2:i386 (1.13.2+dfsg-5ubuntu2) ...
Unpacking libgssapi-krb5-2:amd64 (1.15-1) over (1.13.2+dfsg-5ubuntu2) ...
Preparing to unpack .../libgssapi-krb5-2_1.15-1_i386.deb ...
Unpacking libgssapi-krb5-2:i386 (1.15-1) over (1.13.2+dfsg-5ubuntu2) ...
Preparing to unpack .../libkrb5-3_1.15-1_i386.deb ...
De-configuring libkrb5-3:amd64 (1.13.2+dfsg-5ubuntu2) ...
Unpacking libkrb5-3:i386 (1.15-1) over (1.13.2+dfsg-5ubuntu2) ...
Preparing to unpack .../libkrb5-3_1.15-1_amd64.deb ...
Unpacking libkrb5-3:amd64 (1.15-1) over (1.13.2+dfsg-5ubuntu2) ...
Preparing to unpack .../libkrb5support0_1.15-1_amd64.deb ...
De-configuring libkrb5support0:i386 (1.13.2+dfsg-5ubuntu2) ...
Unpacking libkrb5support0:amd64 (1.15-1) over (1.13.2+dfsg-5ubuntu2) ...
Preparing to unpack .../libkrb5support0_1.15-1_i386.deb ...
Unpacking libkrb5support0:i386 (1.15-1) over (1.13.2+dfsg-5ubuntu2) ...
Preparing to unpack .../libk5crypto3_1.15-1_i386.deb ...
De-configuring libk5crypto3:amd64 (1.13.2+dfsg-5ubuntu2) ...
Unpacking libk5crypto3:i386 (1.15-1) over (1.13.2+dfsg-5ubuntu2) ...
Preparing to unpack .../libk5crypto3_1.15-1_amd64.deb ...
Unpacking libk5crypto3:amd64 (1.15-1) over (1.13.2+dfsg-5ubuntu2) ...
Preparing to unpack .../libproxy1-plugin-webkit_0.4.14-2_i386.deb ...
Unpacking libproxy1-plugin-webkit:i386 (0.4.14-2) over (0.4.11-5ubuntu1) ...
Preparing to unpack .../libproxy1-plugin-gsettings_0.4.14-2_i386.deb ...
De-configuring libproxy1-plugin-gsettings:amd64 (0.4.11-5ubuntu1) ...
Unpacking libproxy1-plugin-gsettings:i386 (0.4.14-2) over (0.4.11-5ubuntu1) ...
Preparing to unpack .../libproxy1-plugin-gsettings_0.4.14-2_amd64.deb ...
Unpacking libproxy1-plugin-gsettings:amd64 (0.4.14-2) over (0.4.11-5ubuntu1) ...
Preparing to unpack .../libproxy1-plugin-networkmanager_0.4.14-2_amd64.deb ...
Unpacking libproxy1-plugin-networkmanager:amd64 (0.4.14-2) over (0.4.11-5ubuntu1) ...
Preparing to unpack .../libproxy1v5_0.4.14-2_amd64.deb ...
De-configuring libproxy1v5:i386 (0.4.11-5ubuntu1) ...
Unpacking libproxy1v5:amd64 (0.4.14-2) over (0.4.11-5ubuntu1) ...
Preparing to unpack .../libproxy1v5_0.4.14-2_i386.deb ...
Unpacking libproxy1v5:i386 (0.4.14-2) over (0.4.11-5ubuntu1) ...
Preparing to unpack .../libopenconnect5_7.08-1_amd64.deb ...
Unpacking libopenconnect5:amd64 (7.08-1) over (7.06-2build2) ...
Processing triggers for libc-bin (2.23-0ubuntu9) ...
Processing triggers for man-db (2.7.5-1) ...
Setting up libtasn1-6:amd64 (4.10-1) ...
Setting up libtasn1-6:i386 (4.10-1) ...
Setting up libtasn1-6-dev:amd64 (4.10-1) ...
Setting up libgnutls30:amd64 (3.5.6-4ubuntu4) ...
Setting up libgnutls30:i386 (3.5.6-4ubuntu4) ...
Setting up libgnutls-openssl27:amd64 (3.5.6-4ubuntu4) ...
Setting up libgnutlsxx28:amd64 (3.5.6-4ubuntu4) ...
Setting up libgnutls28-dev:amd64 (3.5.6-4ubuntu4) ...
Setting up libkrb5support0:amd64 (1.15-1) ...
Setting up libkrb5support0:i386 (1.15-1) ...
Setting up libk5crypto3:amd64 (1.15-1) ...
Setting up libk5crypto3:i386 (1.15-1) ...
Setting up libkrb5-3:amd64 (1.15-1) ...
Setting up libkrb5-3:i386 (1.15-1) ...
Setting up libgssapi-krb5-2:amd64 (1.15-1) ...
Setting up libgssapi-krb5-2:i386 (1.15-1) ...
Setting up libgssrpc4:amd64 (1.15-1) ...
Setting up libkadm5srv-mit11:amd64 (1.15-1) ...
Setting up libkadm5clnt-mit11:amd64 (1.15-1) ...
Setting up krb5-multidev (1.15-1) ...
Setting up libproxy1v5:i386 (0.4.14-2) ...
Setting up libproxy1v5:amd64 (0.4.14-2) ...
Setting up libproxy1-plugin-webkit:i386 (0.4.14-2) ...
Setting up libproxy1-plugin-gsettings:amd64 (0.4.14-2) ...
Setting up libproxy1-plugin-gsettings:i386 (0.4.14-2) ...
Setting up libproxy1-plugin-networkmanager:amd64 (0.4.14-2) ...
Setting up libopenconnect5:amd64 (7.08-1) ...
Processing triggers for libc-bin (2.23-0ubuntu9) ...
$
```
Note that above I only installed libopenconnect5 7.08, so that:
```sh
$ openconnect --version
WARNING: This version of openconnect is v7.06 but
         the libopenconnect library is v7.08
OpenConnect version v7.08
Using GnuTLS. Features present: PKCS#11, RSA software token, HOTP software token, TOTP software token, Yubikey OATH, System keys, DTLS
```
But that seems sufficient for connecting to the VPN;
```sh
```
$ sudo openconnect --authgroup="Single-Factor Pulse Clients" --juniper https://remote.ucsf.edu/openconnect
WARNING: This version of openconnect is v7.06 but
         the libopenconnect library is v7.08
WARNING: Juniper Network Connect support is experimental.
It will probably be superseded by Junos Pulse support.
GET https://remote.ucsf.edu/openconnect
Connected to 169.230.251.90:443
SSL negotiation with remote.ucsf.edu
Connected to HTTPS on remote.ucsf.edu
Got HTTP response: HTTP/1.1 302 Found
GET https://remote.ucsf.edu/dana-na/auth/url_5/welcome.cgi
SSL negotiation with remote.ucsf.edu
Connected to HTTPS on remote.ucsf.edu
*** IMPORTANT UCSF VPN NOTIFICATIONS ***

Remember to use your Active Directory ID (username) when you log into VPN, NOT your MyAccess SFID or UCSF email address. Starting Tuesday 11/28, you can ONLY use your AD ID to log into VPN.

Starting Tuesday 12/5, you will need to use multi-factor authentication (Duo) to log into VPN; sign up for Duo at it.ucsf.edu/duo or contact the Service Desk (help.ucsf.edu or 415-514-4100).

Click 'Proceed' below to continue.
POST https://remote.ucsf.edu/dana-na/auth/url_5/welcome.cgi
SSL negotiation with remote.ucsf.edu
Connected to HTTPS on remote.ucsf.edu
Got HTTP response: HTTP/1.1 302 Moved
GET https://remote.ucsf.edu/dana-na/auth/url_5/welcome.cgi?
SSL negotiation with remote.ucsf.edu
Connected to HTTPS on remote.ucsf.edu
frmLogin
username: <user>
password: <pwd>
POST https://remote.ucsf.edu/dana-na/auth/url_5/login.cgi
SSL negotiation with remote.ucsf.edu
Connected to HTTPS on remote.ucsf.edu
Got HTTP response: HTTP/1.1 302 Moved
GET https://remote.ucsf.edu/dana-na/auth/url_5/welcome.cgi?p=user-confirm&id=state_da1465d08302a674de298490d8f086c7
SSL negotiation with remote.ucsf.edu
Connected to HTTPS on remote.ucsf.edu
POST https://remote.ucsf.edu/dana-na/auth/url_5/login.cgi
SSL negotiation with remote.ucsf.edu
Connected to HTTPS on remote.ucsf.edu
Got HTTP response: HTTP/1.1 302 Moved
GET https://remote.ucsf.edu/dana/home/starter0.cgi?check=yes
SSL negotiation with remote.ucsf.edu
Connected to HTTPS on remote.ucsf.edu
SSL negotiation with remote.ucsf.edu
Connected to HTTPS on remote.ucsf.edu
SSL negotiation with remote.ucsf.edu
Connected to HTTPS on remote.ucsf.edu
ESP session established with server
```

After confirming this, I removed those Zesty entries from `/etc/apt/sources.list` again.

**Comment by HenrikBengtsson** [comment]
If anyone knows a better way to install OpenConnect 7.08 on Ubuntu 16.04, please let me/us know.

**Comment by HenrikBengtsson** [comment]
NOTES: While trying to install 64-bit Pulse Secure Client 5.3-3-Build553 on Ubuntu 16.04 (Xenial) with above OpenConnect 7.08 (17.04 Zesty hack) already on the system, I ended up with the following unsolvable conflict:
```sh
sudo apt-get install libproxy1-plugin-webkit
Reading package lists... Done
Building dependency tree... 50%
Building dependency tree       
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 libproxy1-plugin-webkit : Depends: libproxy1v5 (= 0.4.11-5ubuntu1) but 0.4.14-2 is to be installed
E: Unable to correct problems, you have held broken packages.
```
Note that the dependency is strictly **equal to** and not **greater or equal to**.  libproxy1v5 0.4.11-5ubuntu1 is what [Ubuntu 16.04 provides](https://packages.ubuntu.com/xenial/net/libproxy1v5), whereas that libproxy1v5 0.4.14-2 is from [Ubuntu 17.04](https://packages.ubuntu.com/zesty/net/libproxy1v5).

From this I conclude that you cannot have OpenConnect 7.08 installed at the same time as the 64-bit Pulse Secure Client 5.3-3-Build553 client.

**Comment by HenrikBengtsson** [comment]
@graft has also confirmed that two-factor authentication (2FA) works with OpenConnect 7.08.  He can connect to the UCSF VPN using:
```sh
$ sudo openconnect --authgroup="Dual-Factor Pulse Clients" --juniper https://remote.ucsf.edu/openconnect
```
Comment: URL `https://remote.ucsf.edu/pulse` probably works equally well as `https://remote.ucsf.edu/openconnect` - at least both work with the Pulse Secure client.

**Comment by HenrikBengtsson** [comment]
I can confirm that Ubuntu 18.04 LTS provides OpenConnect 7.08 and it's possible to connect to the UCSF VPN with 2FA _interactively_ at the CLI using:
```sh
sudo openconnect --authgroup="Dual-Factor Pulse Clients" --juniper https://remote.ucsf.edu/openconnect
```
When I get time, I'll update `ucsf vpn` to also support OpenConnect (as it did in the past prior to the major UCSF update).

**Comment by HenrikBengtsson** [comment]
ucsf-vpn 4.0.0 was just released - it uses OpenConnect (>= 7.08) by default, which is now available on Ubuntu 18.04 LTS.  Pulse Secure client is still supported via `--method pulse` (or env var `UCSF_VPN_METHOD=pulse`).

**Comment by graft** [comment]
An issue I have been having lately with openconnect on some of my machines is a strange name service bug - the connection looks fine, an IP is assigned, but there is no ping available. This only happens on systems that use systemd-resolved - other systems I have that still use dnsmasq seem to work okay. I haven't really been able to understand what's going on here, but there is some discussion of the issue and a workaround here: https://github.com/dlenski/openconnect/issues/104

The workaround, in case you encounter this issue on a new install (I'm running 18.04 ubuntu) is, after your connection is active, look for the following process:

    run-parts --arg=-a --arg=tun0 /etc/resolvconf/update.d

Killing this will give you a working VPN connection.
